SQLiteOpenHelperTest causing FileSystemException on Windows

`SQLiteOpenHelperTest` has never run well on Windows, however with a recent change on master, (looks like attempted cleanup code to remove the temp directory) - I have been running into the following:

      Caused by: java.nio.file.FileSystemException:
      C:\Users\Jeremy\AppData\Local\Temp\android-tmp-robolectric250258741663565975\databases\path:
      The process cannot access the file because it is being used by another process.

        at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:86)
        at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
        at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
        at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:269)
        at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
        at java.nio.file.Files.delete(Files.java:1123)
        at org.robolectric.util.TempDirectory$2.visitFile(TempDirectory.java:84)
        at org.robolectric.util.TempDirectory$2.visitFile(TempDirectory.java:81)
        at java.nio.file.Files.walkFileTree(Files.java:2667)
        at java.nio.file.Files.walkFileTree(Files.java:2739)
        at org.robolectric.util.TempDirectory.clearDirectory(TempDirectory.java:81)
        at org.robolectric.util.TempDirectory.destroyImpl(TempDirectory.java:70)
        ... 27 more

What seems to happen is that the tests run fine until it gets to the `SQLiteOpenHelperTest`, and thereafter all the tests fail with the above error. The SQLite stuff has never played nice under Windows - Windows is a a little bit more pedantic about removing files that are still in use, and I've always had errors about a process still having the database open in the SQLite-related tests. As I haven't been doing any testing with SQLite so far I've just been ignoring these errors, but now that it's causing other unrelated tests to fail too it gets a bit harder to ignore it.

I'm wondering if this is related to #1510 - if a static reference is held to the connection that persists between tests, then that would hold the file open (and locked) across tests, which would cause the above.

--- a/robolectric/src/main/java/org/robolectric/Robolectric.java
+++ b/robolectric/src/main/java/org/robolectric/Robolectric.java
@@ -15,19 +15,18 @@ import static org.robolectric.Shadows.shadowOf;
 
 public class Robolectric {
   private static final ShadowsAdapter shadowsAdapter = instantiateShadowsAdapter();
-  private static ServiceLoader<ShadowProvider> providers;
+  private static Iterable<ShadowProvider> providers;
 
   public static void reset() {
-    RuntimeEnvironment.application = null;
-    RuntimeEnvironment.setRobolectricPackageManager(null);
-    RuntimeEnvironment.setActivityThread(null);
-
     if (providers == null) {
       providers = ServiceLoader.load(ShadowProvider.class);
     }
     for (ShadowProvider provider : providers) {
       provider.reset();
     }
+    RuntimeEnvironment.application = null;
+    RuntimeEnvironment.setRobolectricPackageManager(null);
+    RuntimeEnvironment.setActivityThread(null);
   }
 
   public static ShadowsAdapter getShadowsAdapter() {

--- a/robolectric/src/main/java/org/robolectric/internal/ParallelUniverse.java
+++ b/robolectric/src/main/java/org/robolectric/internal/ParallelUniverse.java
@@ -3,25 +3,26 @@ package org.robolectric.internal;
 import android.app.Application;
 import android.content.Context;
 import android.content.pm.ApplicationInfo;
-import android.content.pm.PackageInfo;
 import android.content.pm.PackageManager;
 import android.content.res.Configuration;
 import android.content.res.Resources;
-import java.nio.file.Paths;
+
 import org.bouncycastle.jce.provider.BouncyCastleProvider;
-import org.robolectric.*;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.RuntimeEnvironment;
+import org.robolectric.ShadowsAdapter;
+import org.robolectric.TestLifecycle;
 import org.robolectric.annotation.Config;
 import org.robolectric.internal.fakes.RoboInstrumentation;
 import org.robolectric.manifest.AndroidManifest;
 import org.robolectric.res.ResBunch;
 import org.robolectric.res.ResourceLoader;
 import org.robolectric.res.builder.DefaultPackageManager;
-import org.robolectric.res.builder.RobolectricPackageManager;
 import org.robolectric.util.ReflectionHelpers;
 
 import java.lang.reflect.Method;
 import java.security.Security;
-import org.robolectric.util.TempDirectory;
 
 import static org.robolectric.util.ReflectionHelpers.ClassParameter;
 
@@ -142,13 +143,6 @@ public class ParallelUniverse implements ParallelUniverseInterface {
 
   @Override
   public void tearDownApplication() {
-    // We would like to null this out (which would call reset) but that can kill async threads
-    // that run after a test complete (because they might call Context.getApplicationInfo()).
-    RobolectricPackageManager rpm = RuntimeEnvironment.getRobolectricPackageManager();
-    if (rpm != null) {
-      rpm.reset();
-    }
-
     if (RuntimeEnvironment.application != null) {
       RuntimeEnvironment.application.onTerminate();
     }

--- a/robolectric/src/test/java/org/robolectric/ParallelUniverseTest.java
+++ b/robolectric/src/test/java/org/robolectric/ParallelUniverseTest.java
@@ -7,16 +7,20 @@ import org.junit.Test;
 import org.junit.Before;
 import org.junit.runner.RunWith;
 import org.junit.runners.model.InitializationError;
-import org.robolectric.annotation.Config;
+
 import android.content.res.Resources;
 import android.content.res.Configuration;
+
+import org.robolectric.annotation.Config;
 import org.robolectric.internal.ParallelUniverse;
 import org.robolectric.internal.SdkConfig;
+import org.robolectric.res.builder.RobolectricPackageManager;
 
 import java.security.cert.CertificateException;
 import java.security.cert.CertificateFactory;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
 
 @RunWith(TestRunners.WithDefaults.class)
 public class ParallelUniverseTest {
@@ -63,6 +67,21 @@ public class ParallelUniverseTest {
     assertThat(getQualifiersfromSystemResources()).isEqualTo("large-land-v18");
     assertThat(getQualifiersFromAppAssetManager()).isEqualTo("large-land-v18");
     assertThat(getQualifiersFromSystemAssetManager()).isEqualTo("large-land-v18");
+  }
+  
+  @Test
+  public void tearDownApplication_shouldNotResetPackageManager() {
+    RobolectricPackageManager pm = mock(RobolectricPackageManager.class);
+    RuntimeEnvironment.setRobolectricPackageManager(pm);
+    pu.tearDownApplication();
+    verify(pm, never()).reset();
+  }
+  
+  @Test
+  public void tearDownApplication_invokesOnTerminate() {
+    RuntimeEnvironment.application = mock(Application.class);
+    pu.tearDownApplication();
+    verify(RuntimeEnvironment.application).onTerminate();
   }
   
   private String getQualifiersfromSystemResources() {

--- a/robolectric/src/test/java/org/robolectric/RobolectricTest.java
+++ b/robolectric/src/test/java/org/robolectric/RobolectricTest.java
@@ -8,26 +8,36 @@ import android.os.Bundle;
 import android.os.Handler;
 import android.view.Display;
 import android.view.View;
+
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
 import org.robolectric.annotation.Implementation;
 import org.robolectric.annotation.Implements;
+import org.robolectric.res.builder.RobolectricPackageManager;
 import org.robolectric.shadows.ShadowApplication;
 import org.robolectric.shadows.ShadowDisplay;
 import org.robolectric.shadows.ShadowLooper;
 import org.robolectric.shadows.ShadowView;
 import org.robolectric.shadows.StubViewRoot;
 import org.robolectric.internal.Shadow;
+import org.robolectric.internal.ShadowProvider;
+import org.robolectric.util.ReflectionHelpers;
 import org.robolectric.util.TestOnClickListener;
 
 import java.io.ByteArrayOutputStream;
 import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.*;
 import static org.robolectric.Shadows.shadowOf;
 
 @RunWith(TestRunners.WithDefaults.class)
@@ -138,9 +148,56 @@ public class RobolectricTest {
     assertThat(activity.isVisible()).isTrue();
   }
 
+  private List<String> order = new ArrayList<>();
+  
+  private class MockProvider implements ShadowProvider {
+    @Override
+    public void reset() {
+      order.add("shadowProvider");
+      assertThat(RuntimeEnvironment.application).as("app during shadow reset").isNotNull();
+      assertThat(RuntimeEnvironment.getActivityThread()).as("activityThread during shadow reset").isNotNull();
+      assertThat(RuntimeEnvironment.getRobolectricPackageManager()).as("packageManager during shadow reset").isNotNull();
+    }
+
+    @Override
+    public String[] getProvidedPackageNames() {
+      return null;
+    }
+  }
+
+  @Test
+  public void reset_shouldResetShadows_beforeClearingPackageManager() {
+    Iterable<ShadowProvider> oldProviders = ReflectionHelpers.getStaticField(Robolectric.class, "providers");;
+    ShadowProvider mockProvider = new MockProvider();
+    List<ShadowProvider> mockProviders = Collections.singletonList(mockProvider);
+    
+    ReflectionHelpers.setStaticField(Robolectric.class, "providers", mockProviders);
+
+    RobolectricPackageManager mockManager = mock(RobolectricPackageManager.class);
+    doAnswer(new Answer<Void>() {
+      public Void answer(InvocationOnMock invocation) {
+        order.add("packageManager");
+        return null;
+      }
+    }).when(mockManager).reset();
+    
+    RuntimeEnvironment.setRobolectricPackageManager(mockManager);
+    
+    try {
+      Robolectric.reset();
+      
+    } finally {
+      // Make sure we clean up after ourselves
+      ReflectionHelpers.setStaticField(Robolectric.class, "providers", oldProviders);
+    }
+    assertThat(order).as("reset order").containsExactly("shadowProvider", "packageManager");
+    assertThat(RuntimeEnvironment.application).as("app after reset").isNull();
+    assertThat(RuntimeEnvironment.getPackageManager()).as("packageManager after reset").isNull();
+    assertThat(RuntimeEnvironment.getActivityThread()).as("activityThread after reset").isNull();
+  }
+  
   @Implements(View.class)
   public static class TestShadowView {
-    @SuppressWarnings({"UnusedDeclaration"})
     @Implementation
     public Context getContext() {
       return null;

--- a/robolectric/src/test/java/org/robolectric/RobolectricTestRunnerSelfTest.java
+++ b/robolectric/src/test/java/org/robolectric/RobolectricTestRunnerSelfTest.java
@@ -2,37 +2,52 @@ package org.robolectric;
 
 import android.app.Application;
 import android.content.res.Resources;
+
+import org.assertj.core.api.Assertions;
+import org.junit.AfterClass;
+import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.model.InitializationError;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
 import org.robolectric.annotation.Config;
 import org.robolectric.manifest.AndroidManifest;
+import org.robolectric.res.builder.RobolectricPackageManager;
 import org.robolectric.shadows.ShadowApplication;
 import org.robolectric.util.ReflectionHelpers;
-import java.lang.reflect.Method;
 
-import static org.junit.Assert.*;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
 
 @RunWith(RobolectricTestRunnerSelfTest.RunnerForTesting.class)
 public class RobolectricTestRunnerSelfTest {
 
   @Test
-  public void shouldInitializeAndBindApplicationButNotCallOnCreate() throws Exception {
-    assertNotNull(RuntimeEnvironment.application);
-    assertEquals(MyTestApplication.class, RuntimeEnvironment.application.getClass());
-    assertTrue(((MyTestApplication) RuntimeEnvironment.application).onCreateWasCalled);
-    assertNotNull(ShadowApplication.getInstance().getResourceLoader());
+  public void shouldInitializeAndBindApplicationButNotCallOnCreate() {
+    assertThat(RuntimeEnvironment.application).as("application")
+      .isNotNull()
+      .isInstanceOf(MyTestApplication.class);
+    assertThat(((MyTestApplication) RuntimeEnvironment.application).onCreateWasCalled).as("onCreate called").isTrue();
+    assertThat(ShadowApplication.getInstance().getResourceLoader()).as("resource loader").isNotNull();
   }
 
-  @Test public void shouldSetUpSystemResources() throws Exception {
-    assertNotNull(Resources.getSystem());
-    assertEquals(RuntimeEnvironment.application.getResources().getString(android.R.string.copy),
-        Resources.getSystem().getString(android.R.string.copy));
+  @Test
+  public void shouldSetUpSystemResources() {
+    assertThat(Resources.getSystem()).as("system resources").isNotNull();
+    assertThat(Resources.getSystem().getString(android.R.string.copy)).as("system resource")
+      .isEqualTo(RuntimeEnvironment.application.getResources().getString(android.R.string.copy));
 
-    assertNotNull(RuntimeEnvironment.application.getResources().getString(R.string.howdy));
+    assertThat(RuntimeEnvironment.application.getResources().getString(R.string.howdy)).as("app resource")
+      .isNotNull();
     try {
       Resources.getSystem().getString(R.string.howdy);
-      fail("should have thrown");
+      Assertions.failBecauseExceptionWasNotThrown(Resources.NotFoundException.class);
     } catch (Resources.NotFoundException e) {
     }
   }
@@ -41,26 +56,45 @@ public class RobolectricTestRunnerSelfTest {
   public void setStaticValue_shouldIgnoreFinalModifier() {
     ReflectionHelpers.setStaticField(android.os.Build.class, "MODEL", "expected value");
 
-    assertEquals("expected value", android.os.Build.MODEL);
+    assertThat(android.os.Build.MODEL).isEqualTo("expected value");
   }
 
   @Test
   @Config(qualifiers = "fr")
   public void internalBeforeTest_testValuesResQualifiers() {
     String expectedQualifiers = "fr" + TestRunners.WithDefaults.SDK_TARGETED_BY_MANIFEST;
-    assertEquals(expectedQualifiers, Shadows.shadowOf(ShadowApplication.getInstance().getResources().getAssets()).getQualifiers());
+    assertThat(Shadows.shadowOf(ShadowApplication.getInstance().getResources().getAssets()).getQualifiers()).isEqualTo(expectedQualifiers);
   }
 
   @Test
   public void internalBeforeTest_resetsValuesResQualifiers() {
-    assertEquals("", Shadows.shadowOf(ShadowApplication.getInstance().getResources().getConfiguration()).getQualifiers());
+    assertThat(Shadows.shadowOf(ShadowApplication.getInstance().getResources().getConfiguration()).getQualifiers())
+      .isEqualTo("");
   }
 
   @Test
   public void internalBeforeTest_doesNotSetI18nStrictModeFromSystemIfPropertyAbsent() {
-    assertFalse(ShadowApplication.getInstance().isStrictI18n());
+    assertThat(ShadowApplication.getInstance().isStrictI18n()).isFalse();
   }
 
+  @Before
+  public void clearOrder() {
+    order.clear();
+    RobolectricPackageManager mockManager = mock(RobolectricPackageManager.class);
+    doAnswer(new Answer<Void>() {
+      public Void answer(InvocationOnMock invocation) {
+        order.add("reset");
+        return null;
+      }
+    }).when(mockManager).reset();
+    
+    RuntimeEnvironment.setRobolectricPackageManager(mockManager);
+  }
+  @AfterClass
+  public static void resetStaticState_shouldBeCalled_afterAppTearDown() throws InitializationError {
+    assertThat(order).containsExactly("onTerminate", "reset");
+  }
+    
   public static class RunnerForTesting extends TestRunners.WithDefaults {
     public static RunnerForTesting instance;
 
@@ -80,11 +114,19 @@ public class RobolectricTestRunnerSelfTest {
     }
   }
 
+  private static List<String> order = new ArrayList<>();
+  
   public static class MyTestApplication extends Application {
     private boolean onCreateWasCalled;
 
-    @Override public void onCreate() {
+    @Override
+    public void onCreate() {
       this.onCreateWasCalled = true;
+    }
+    
+    @Override
+    public void onTerminate() {
+      order.add("onTerminate");
     }
   }
 }

